### ISSUE TITLE
Mesh element number in parallel mesh

### ISSUE THREAD

**Issue by wcdawn** [mfem/mfem#2416]
I have a separate program that calculates the power in each element of a mesh and I'd like to use that in a coefficient for the heat conduction equation. Basically, I have a vector of length `nElement` so I'd like to have a coefficient `return vec(T.ElementNo)` (for `ElementTransformation &T`).

In serial, this works fine.
In parallel, the element numbering is different. I tried using `ParMesh::GetGlobalElementNum()` and it doesn't work as I'd expect.

To look at this, I wrote a coefficient to simply return the "global" element number.
```cpp
double MyCoefficient::Eval(ElementTransformation &T, const IntegrationPoint &ip)
{
  // pmesh is a private member pointer to the ParMesh being used
  return pmesh->GetGlobalElementNum(T.ElementNo);
}
```

For one MPI process, everything works as expected. Here is a plot of a clip of the spherical mesh I'm using.
![element_np1](https://user-images.githubusercontent.com/20099247/126360839-5f3e539d-5cec-41c3-91ba-5eedf2a250cf.png)

For two MPI processes, the "global" element number is different. I would expect both of these plots to be the same.
![element_np2](https://user-images.githubusercontent.com/20099247/126360858-19df8f66-bf1f-4289-9fc6-9eb0c796bdee.png)

Basically, I need to know the element number from the serial mesh given the element number in the parallel mesh.
I tried reading #2311 but it didn't seem to solve my problem.

P.S. Yes, for now I'm storing the whole power distribution (length `nElement`) on each MPI process. It is small for now. Once I figure out how to index it, I'll look into partitioning it.

Thanks!

**Comment by tzanio** [comment]
@wcdawn, to clarify -- Do you want to transfer values associated with each element when transitioning from serial `Mesh` to parallel `ParMesh`?

This has been discussed before, and is not trivial, because in parallel the numbering of the elements is different, for example the elements belonging to an MPI rank have consecutive numbers.

One way you can achieve that is to define L2 grid function of order 0 and transfer it with this method:

https://github.com/mfem/mfem/blob/8923656d477be5f42271af92e5c4eb5008c947f0/fem/pgridfunc.hpp#L78-L82

See e.g. https://github.com/mfem/mfem/issues/2340.

Hope this helps,
Tzanio

**Comment by wcdawn** [comment]
This seems like a good option. How can I obtain the `partitioning` array? I'm using METIS so generally I don't know how the mesh is partitioned.

**Comment by tzanio** [comment]
You can generate the partitioning first, save it, provide it as the optional argument to the `ParMesh` constructor, and then use it in the above method.

Something like:

```c++
int part_method = 1; // METIS options to use, see Mesh::GeneratePartitioning

int *partitioning = mesh.GeneratePartitioning(NRanks, part_method);

ParMesh pmesh(MPI_COMM_WORLD, mesh, partitioning);

ParGridFunction pgf(pmesh, gf, partitioning); 
```

**Comment by wcdawn** [comment]
This Works great! Thanks!
